src/internal/NeoEsp8266AsyncUartMethod.h

```diff
@@ -0,0 +1,89 @@
+/*-------------------------------------------------------------------------
+NeoPixel library helper functions for using asynchronous UART1 in Esp8266.
+
+Written by Unai Uribarri
+
+I invest time and resources providing this open source code,
+please support me by dontating (see https://github.com/Makuna/NeoPixelBus)
+
+-------------------------------------------------------------------------
+This file is part of the Makuna/NeoPixelBus library.
+
+NeoPixelBus is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as
+published by the Free Software Foundation, either version 3 of
+the License, or (at your option) any later version.
+
+NeoPixelBus is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with NeoPixel.  If not, see
+<http://www.gnu.org/licenses/>.
+-------------------------------------------------------------------------*/
+
+#pragma once
+
+#ifdef ARDUINO_ARCH_ESP8266
+#include "NeoEsp8266UartMethod.h"
+
+// due to linker overriding ICACHE_RAM_ATTR for cpp files, this function was moved into a NeoPixelEsp8266-AsyncUart.c file.
+extern "C" {
+    void ICACHE_RAM_ATTR esp8266_uart1_async_prepare();
+    void ICACHE_RAM_ATTR esp8266_uart1_async_send_pixels(uint8_t* pixels, uint8_t* end);
+}
+
+template<typename T_SPEED> class NeoEsp8266AsyncUartMethodBase: public NeoEsp8266UartMethodBase<T_SPEED>
+{
+public:
+    NeoEsp8266AsyncUartMethodBase(uint8_t pin, uint16_t pixelCount, size_t elementSize)
+	: NeoEsp8266UartMethodBase<T_SPEED>(pin, pixelCount, elementSize)
+    {
+        _asyncPixels = (uint8_t*)malloc(this->_sizePixels);
+    }
+
+    ~NeoEsp8266AsyncUartMethodBase()
+    {
+        while (!this->IsReadyToUpdate())
```

Comment:
It just waits until some microseconds until all the asyncPixels data is
written to the UART. This way we output the complete scene before freeing
the allocated data.
On Jun 5, 2016 08:09, "Daniel Bos" <notifications@github.com> wrote:

> In src/internal/NeoEsp8266AsyncUartMethod.h
> <https://github.com/Makuna/NeoPixelBus/pull/108#discussion_r65811818>:
>
> > +    void ICACHE_RAM_ATTR esp8266_uart1_async_prepare();
> > +    void ICACHE_RAM_ATTR esp8266_uart1_async_send_pixels(uint8_t* pixels, uint8_t* end);
> > +}
> > +
> > +template<typename T_SPEED> class NeoEsp8266AsyncUartMethodBase: public NeoEsp8266UartMethodBase<T_SPEED>
> > +{
> > +public:
> > +    NeoEsp8266AsyncUartMethodBase(uint8_t pin, uint16_t pixelCount, size_t elementSize)
> > +	: NeoEsp8266UartMethodBase<T_SPEED>(pin, pixelCount, elementSize)
> > +    {
> > +        _asyncPixels = (uint8_t*)malloc(this->_sizePixels);
> > +    }
> > +
> > +    ~NeoEsp8266AsyncUartMethodBase()
> > +    {
> > +        while (!this->IsReadyToUpdate())
>
> Is this necessary in the destructor? Seems to me this doesn't *"do"*
> anything besides waiting.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Makuna/NeoPixelBus/pull/108/files/1564354eb8f5376e98b74a7849ed257d497d187c#r65811818>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFnLB3Pyyyh2kiUt2lgVQhJUUETJPqPeks5qImgigaJpZM4IuMAL>
> .
>